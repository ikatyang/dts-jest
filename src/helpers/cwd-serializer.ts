export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
