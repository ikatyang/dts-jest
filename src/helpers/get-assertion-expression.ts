import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
