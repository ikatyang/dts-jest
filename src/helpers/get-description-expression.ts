import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
