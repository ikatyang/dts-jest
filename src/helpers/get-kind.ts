import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
