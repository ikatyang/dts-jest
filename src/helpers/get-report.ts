import {indent} from '../utils/indent';

// tslint:disable:no-magic-numbers

export const get_report = (reporter_template: string, expression: string, snapshot: string) =>
  reporter_template
    .split(/\{\{(expression|snapshot),([0-9]+)\}\}/g)
    .map((chunk, index, chunks) =>
      (index % 3 === 0)
          ? chunk
          : (index % 3 === 1)
            ? ''
            : (chunks[index - 1] === 'snapshot')
              ? indent(snapshot, +chunk)
              : indent(expression, +chunk),
    )
    .join('');
