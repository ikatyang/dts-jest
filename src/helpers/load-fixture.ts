import * as fs from 'fs';
import * as path from 'path';
import * as _ts from 'typescript';
import { create_source_file } from '../utils/create-source-file';

export const get_fixture_filename = (id: string) =>
  path.resolve(__dirname, '../../fixtures', id);

export const load_fixture = (id: string) =>
  fs.readFileSync(get_fixture_filename(id), 'utf8');

export const load_fixture_source_file = (id: string, ts: typeof _ts) => {
  const source = load_fixture(id);
  return create_source_file(id, source, ts);
};
