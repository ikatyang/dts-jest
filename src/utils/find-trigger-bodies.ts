import * as _ts from 'typescript';
import { TriggerBody, TriggerGroup, TriggerHeader } from '../definitions';
import { create_message } from './create-message';
import { get_display_line } from './get-display-line';
import { get_node_one_line_text } from './get-node-one-line-text';
import { get_trigger_groups } from './get-trigger-groups';
import { get_trigger_header_line } from './get-trigger-line';
import { traverse_node } from './traverse-node';

export const find_trigger_bodies = (
  source_file: _ts.SourceFile,
  headers: TriggerHeader[],
  ts: typeof _ts,
) => {
  const bodies: TriggerBody[] = [];

  const header_map = new Map<number, TriggerHeader>();
  headers.forEach(header => header_map.set(header.line, header));

  const groups = get_trigger_groups(headers);
  const invalid_groups: TriggerGroup[] = [];

  traverse_node(
    source_file,
    node => {
      const start = node.getStart();
      const end = node.getEnd();

      const body_line = source_file.getLineAndCharacterOfPosition(start).line;
      const body_end_line = source_file.getLineAndCharacterOfPosition(end).line;

      // check if group position is valid (top-level comment)
      if (groups.length !== 0) {
        const [first_group] = groups;

        if (first_group.line < body_line) {
          // checked
          groups.shift();
        } else if (
          // group is not at the top-level (surrounded by node)
          first_group.line >= body_line &&
          first_group.line <= body_end_line
        ) {
          invalid_groups.push(groups.shift()!);
        }
      }

      const header_line = get_trigger_header_line(body_line);

      if (
        node.kind <= ts.SyntaxKind.LastTriviaToken ||
        !header_map.has(header_line)
      ) {
        return;
      }

      header_map.delete(header_line);

      bodies.push({
        start,
        end,
        experssion: get_node_one_line_text(node, source_file, ts),
        text: get_dedented_expression_text(node, source_file)
          // remove trailing semicolons and spaces
          .replace(/\s*;*\s*$/, ''),
      });
    },
    ts,
  );

  // unattachable_lines
  if (header_map.size !== 0) {
    const unattachable_line_infos: string[] = [];
    header_map.forEach((header, line) => {
      const { description = '' } = header;
      unattachable_line_infos.push(
        `${source_file.fileName}:${get_display_line(line)} ${description}`,
      );
    });
    throw new Error(
      create_message(
        'Unattachable trigger(s) detected:',
        unattachable_line_infos,
      ),
    );
  }

  if (invalid_groups.length !== 0) {
    throw new Error(
      create_message(
        'Invalid trigger-group(s) detected:',
        invalid_groups.map(
          ({ line, description = '' }) =>
            `${source_file.fileName}:${get_display_line(line)} ${description}`,
        ),
      ),
    );
  }

  return bodies;
};

function get_dedented_expression_text(
  node: _ts.Node,
  source_file: _ts.SourceFile,
) {
  const start = node.getStart();
  const { character } = source_file.getLineAndCharacterOfPosition(start);

  return (
    node
      .getText(source_file)
      // dedent
      .replace(/^ */gm, spaces =>
        ' '.repeat(Math.max(0, spaces.length - character)),
      )
  );
}
