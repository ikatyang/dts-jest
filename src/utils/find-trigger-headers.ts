import * as _ts from 'typescript';
import {
  trigger_header_regex,
  GroupMethod,
  TestMethod,
  TriggerGroup,
  TriggerHeader,
  TriggerHeaderFlags,
  TriggerHeaderMatchIndex,
} from '../definitions';
import { for_each_comment } from './for-each-comment';
import { get_comment_content } from './get-comment-content';
import { parse_trigger_header_flags } from './parse-trigger-header-flags';

export const find_trigger_headers = (
  source_file: _ts.SourceFile,
  ts: typeof _ts,
) => {
  const headers: TriggerHeader[] = [];

  let last_group: TriggerGroup | undefined;

  for_each_comment(
    source_file,
    (comment, scanner) => {
      const match = get_comment_content(comment).match(trigger_header_regex);

      if (match === null) {
        return;
      }

      const description = match[TriggerHeaderMatchIndex.Description];
      const flags = parse_trigger_header_flags(
        match[TriggerHeaderMatchIndex.Flags],
      );

      const start = scanner.getTokenPos();
      const { line } = source_file.getLineAndCharacterOfPosition(start);

      if (flags & TriggerHeaderFlags[':group']) {
        last_group = {
          line,
          method: get_group_method(flags),
          description: description.length === 0 ? undefined : description,
        };
      } else {
        headers.push({
          line,
          flags,
          method: get_test_method(flags),
          description: description.length === 0 ? undefined : description,
          group: last_group,
        });
      }
    },
    ts,
  );

  return headers;
};

function get_test_method(flag: TriggerHeaderFlags) {
  return flag & TriggerHeaderFlags[':only']
    ? TestMethod.Only
    : flag & TriggerHeaderFlags[':skip'] ? TestMethod.Skip : TestMethod.Test;
}

function get_group_method(flag: TriggerHeaderFlags) {
  return flag & TriggerHeaderFlags[':only']
    ? GroupMethod.Only
    : flag & TriggerHeaderFlags[':skip'] ? GroupMethod.Skip : GroupMethod.Test;
}
