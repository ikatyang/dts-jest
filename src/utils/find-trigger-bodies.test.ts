import * as ts from 'typescript';
import { load_fixture_source_file } from '../helpers/load-fixture';
import { find_trigger_bodies } from './find-trigger-bodies';
import { find_trigger_headers } from './find-trigger-headers';

it('should return correctly', () => {
  expect(find_bodies('simple')).toMatchSnapshot();
});

it('should throw error if there are some unattachable bodies', () => {
  expect(() => find_bodies('unattachable')).toThrowErrorMatchingSnapshot();
});

it('should throw error if there are some invalid groups', () => {
  expect(() => find_bodies('invalid-group')).toThrowErrorMatchingSnapshot();
});

function find_bodies(id: string) {
  const source_file = load_fixture_source_file(
    `find-trigger-bodies/${id}.ts`,
    ts,
  );
  const headers = find_trigger_headers(source_file, ts);
  return find_trigger_bodies(source_file, headers, ts);
}
