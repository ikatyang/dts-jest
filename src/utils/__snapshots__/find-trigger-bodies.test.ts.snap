// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `
Array [
  Object {
    "end": 31,
    "expression": "Math.max(1, 2, 3)",
    "one_line_expression": "Math.max(1, 2, 3)",
    "start": 13,
  },
  Object {
    "end": 73,
    "expression": "Math.max(
  1,
  2,
  3,
)",
    "one_line_expression": "Math.max(1, 2, 3)",
    "start": 46,
  },
  Object {
    "end": 119,
    "expression": "Math.max(1, 2, 3)",
    "one_line_expression": "Math.max(1, 2, 3)",
    "start": 101,
  },
  Object {
    "end": 187,
    "expression": "Math.max(
  1,
  2,
  3,
)",
    "one_line_expression": "Math.max(1, 2, 3)",
    "start": 152,
  },
]
`;

exports[`should throw error if there are some invalid groups 1`] = `
"Invalid trigger-group(s) detected:

  find-trigger-bodies/invalid-group.ts:2
  find-trigger-bodies/invalid-group.ts:14 inside"
`;

exports[`should throw error if there are some unattachable bodies 1`] = `
"Unattachable trigger(s) detected:

  find-trigger-bodies/unattachable.ts:1
  find-trigger-bodies/unattachable.ts:2 description"
`;
